### ISSUE TITLE
[SERVER] Inconsistencies in USASK and BAS mirror 

### ISSUE THREAD

**Issue by mts299** [SuperDARN/DDWG#18]
# Affiliated Servers
- [X] BAS
- [X] USASK
- [ ] VT (holding institute)

## Description 
Hi all,

As some of you have noticed, Evan and I have recently been transferring
a number of dat and rawacf files from Dartmouth to BAS, in an attempt to
reduce the long standing gaps in the BAS mirror coverage.  In particular
rawacf data from 2016, but other gaps also.

In addition, I've also processed the latest USask Globus blacklisted
files list against our mirror.

There are still some differences between the USask Globus mirror and the
BAS mirror, and we are currently trying to characterise these.

Thanks,
Paul & Evan.

**Comment by mts299** [comment]
Email: 2019-11-12

Hello all,

I'm not sure if the hashes on the mirrors have been kept up-to-date, but I compared all of the dat files listed in the BAS hashes to the dat files listed in the Globus hashes (from the all_hashes.tar.bz2 file - the hash files in the individual year/momth directories don't seem to work for me) and the results can be seen in the two attached files (>21,000 in the missing_from_bas.txt and >16,000 in the missing_from_globus.txt files, respectively).

I should point out that the Globus dat hashes for 199810, 199812, 200207, and 200304 are still missing (see my previous email from Sep 3rd), so I was unable to compare those four months.  It also looks like all but one of the dat files missing from the Globus hashes are SH files from 1999, which I thought were recovered earlier this year.

If the hash files on each mirror aren't actually accurate then all of this can be ignored.  But if so, I would kindly ask Kevin and Paul to update those files so we can get a better handle on any inconsistencies between the two mirrors.

Cheers,
Evan

---


Hello all,

Following up on this, I have gone back to Globus and downloaded all of the monthly hash files from within the year/month directories rather than relying upon the almost certainly out-of-date all_hashes.tar.gz file and re-ran my comparison script for both dat and rawacf files from 1993-2018.

The first two attached files ("missing_from_bas_dat.txt" and "missing_from_globus_dat.txt") list the dat files missing from one mirror but present on the other (according to the monthly hashes).  Likewise, the second two attached files ("missing_from_bas_raw.txt" and "missing_from_globus_raw.txt") list the rawacf files missing from one mirror but present on the other.

Again, I think there are issues with the monthly hash files on the Globus mirror.  For example, the dat 200512.hashes only lists 7 dat files when clearly there are thousands of dat files for that month present on Globus.  Even still, there are more than 25,000 rawacf files seemingly missing from the BAS mirror and more than 17,000 rawacf files seemingly missing from the Globus mirror.

Kevin, Paul - any thoughts on what might be happening here?

Cheers,
Evan

---

Email: 2019-11-18
Hello all,

I have refined my script to better handle the cases where sub-2-hour dat filenames with capital letters (eg 2003123123hA.dat.bz2) in the Globus hashes are compared case-insensitively to the dat filenames in the BAS hashes.  Running this comparison between mirror hash files again for 1993-2018, here are the total number of seemingly missing files for each type:

BAS:
- 4,700 dat files
- 25,895 rawacf files

Globus:
- 355,748 dat files
- 17,149 rawacf files

Not to sound alarmist, but that's the equivalent of 934 days (or 2.5 years) of data from 36 radars either missing or unaccounted for in the monthly hash files.  Any clarification here would be greatly appreciated.

Thank you,
Evan

**Comment by mts299** [comment]
Email: 2019-11-18

Hey Evan & DDWG,

Indeed that tar.bz2 file is not kept up to date, I made it at some point on request and didn't remove it. It's now removed.

I have recalculated all the dat hashes files now, and found that the issue occurred where they were overwritten when I added dat files onto the mirror in August.

Kevin K.

---

Hi Evan & DDWG,

My apologies for not noticing the dat hash files issue.
All the files listed in `missing_from_globus_dat.txt` are on the globus mirror.

We are still missing the 2194 dat files remaining from the data loss in 2018.
Again, does anyone in the DDWG think they may have these files somewhere? I've attached the listing.

The ~17,000 files in 'missing_from_globus_raw.txt' are all either blacklisted files or failed files.

You can see the list of blacklisted files on globus at /sddata/config/blacklist/ and you can see the list of failed files at /sddata/config/failed/. These directories have been available at /sddata/.config/ previously, but that's hidden by default on the Globus UI so I've made a softlink to the hidden directory now to make it easier to access. I've also made softlinks to all the hash files in here for ease of downloading them all at once.
** NOTE that you have to drill down into the dat_hashes and raw_hashes directories and select the files themselves for transfer. Globus won't follow the symlinks if you just select the directory for transfer. **

In my most recent round of 'gaps checking' I've sent PIs/data managers their list of failed files to see if they have any good versions.
Thank you to those who've responded so far.

Regards,
Kevin

**Comment by mts299** [comment]
Email: 2019-11-18

Hi Evan & DDWG,

My apologies for not noticing the dat hash files issue.
All the files listed in `missing_from_globus_dat.txt` are on the globus mirror.

We are still missing the 2194 dat files remaining from the data loss in 2018.
Again, does anyone in the DDWG think they may have these files somewhere? I've attached the listing.

The ~17,000 files in 'missing_from_globus_raw.txt' are all either blacklisted files or failed files.

You can see the list of blacklisted files on globus at /sddata/config/blacklist/ and you can see the list of failed files at /sddata/config/failed/. These directories have been available at /sddata/.config/ previously, but that's hidden by default on the Globus UI so I've made a softlink to the hidden directory now to make it easier to access. I've also made softlinks to all the hash files in here for ease of downloading them all at once.
** NOTE that you have to drill down into the dat_hashes and raw_hashes directories and select the files themselves for transfer. Globus won't follow the symlinks if you just select the directory for transfer. **

In my most recent round of 'gaps checking' I've sent PIs/data managers their list of failed files to see if they have any good versions.
Thank you to those who've responded so far.

Regards,
Kevin

---

Hi Kevin,

Thanks for recalculating the monthly hashes for the dat files.  I'm no longer finding any dat files present on the BAS mirror but missing from Globus.  On the other hand, the number of dat files missing from BAS has increased from 4,700 at last count to 16,844 (see attached file).

I'm also still puzzled by the dat files with upper-case letters used to indicate the sub-2-hour files, as I don't think that matches the standard file naming scheme.  Were those the recently recovered SH files?

Evan

---

Hi Kevin,

Thanks for looking into this and changing the visibility of that hidden directory on the Globus UI where the blacklisted and failed file lists are located.

So is it correct that any blacklisted or failed files are excluded from the monthly hash files on Globus?  Because they do seem to be included in the BAS hash files (leading to many of the discrepancies noted in my previous emails).

Also, would it be possible to stage up the ~4,700 dat files and ~26,000 rawacf files available on Globus but not the BAS mirror for Paul to download?

Thank you,
Evan

**Comment by mts299** [comment]
Email: 2019-12-03

Hi all,

Just to follow up on this, Kevin - would it be possible to stage up the ~4,700 dat files and ~26,000 rawacf files available on Globus but not the BAS mirror for Paul to download?

Also, any comment from Kevin or Paul on the discrepancies between the two mirrors when including/excluding blacklisted or failed files in the monthly hash files?  Should we settle on one set of rules for generating the hash files?

Thank you,
Evan

**Comment by mts299** [comment]
Email 2019-12-04

Hi Evan, Gareth & DDWG,

Gareth, if Paul has no time to dedicate to the BAS mirror, are you able to find someone else who can spend a bit of time on it?

My comment regarding the hashes files: They are intended to contain only the files that are 'good', i.e. not in the blacklist, or the failed files list.
I believe the hashes files on the BAS mirror have blacklisted files in them that need to be removed from the hash file. I believe Paul knows this.

Regarding the failed files, I sent Paul the 'backscatter' code along with the block of code to use it in May. The link to get the code package is here: https://github.com/SuperDARNCanada/backscatter
I don't think he's had time to implement it, since many failed files still appear on the BAS mirror. For completeness, here it is again:
> Here's the package written by Keith Kotyk here at usask: 
> https://github.com/SuperDARNCanada/backscatter
> And here's the code you need in python to make this check:
>
> from backscatter import dmap
>
> try:
>                 dmap.parse_dmap_format_from_file("filename",raw_dmap=True)
> except dmap.DmapDataError as e:
>                 print("Error. File {0} failed with error 
> {1}".format(filename, e))
> else:
>                 print("{0} passed dmap tests.".format(datafile))
>
> Kevin
I do think that he has the bzip2 tests implemented however, from the very beginning, as well as file size checking (check for 0 size, or size 14 if bzipped).


Regarding Globus and the files that BAS is missing:

Paul (and anyone else) can access the files he's missing via Globus. He'll need to make a user account here: https://www.globus.org/
Then tell me his username so I can share the endpoint with him. This should take about 5 minutes.

Then he (or others) needs to install globusconnectpersonal on his server, from here: https://www.globus.org/globus-connect-personal and specifically for linux: https://docs.globus.org/how-to/globus-connect-personal-linux/
This should take about 10-15 minutes.

At that point, he (or others) can freely use this sync script: https://github.com/SuperDARNCanada/globus/
Which can be automated via cron after running it the first time manually (see instructions at the top of the script, or in the github readme file, which is right here: https://github.com/SuperDARNCanada/globus)

I'm willing to help whomever at BAS set this up so data distribution can be more efficient in the future.

Regards,
Kevin Krieger

**Comment by mts299** [comment]
Email 2020-01-11

Hi Evan, Kevin,

Apologies for the lateness of my reply.

I've regenerated all the dat hashes on our mirror, and when comparing to
the (sorted) original dat hashes, they are identical.  Hence those files
must be genuinely missing on the BAS mirror.

Kevin: I think these missing files are in addition to the ones that you
identified in a previous email, as they were rawacf files.  That's
right, isn't it?

Thanks,
Paul.

**Comment by mts299** [comment]
Email 2020-01-21

Hi Paul,

Thanks for following up on this.  It's been a while since I've looked at this issue, but please let me know if there's anything I can do to help.

Cheers,
Evan

**Comment by mts299** [comment]
Email 2020-01-22

Hi Evan,

As a follow up to this, and from a long-standing issue that our mirror
has had, I've removed all entries from the monthly hash files for files
that have been blacklisted.  This is something that Kevin K. and I
discussed ages ago, but I've only got around to doing recently.

Furthermore, from now on when a file is blacklisted, its entry is
automatically removed from the corresponding monthly hash file.  This
should help future hash file comparisons, but unfortunately won't help
in this particular case.

Cheers,
Paul.
